### DISCUSSION TITLE
NextJS as the front-end framework

### DISCUSSION THREAD

**Discussion by oliverroick** [developmentseed/bioacoustics-frontend#1] [Decision Records]
## Context

To deliver the project on time and on budget, we are looking to use a framework to build the front-end side of the application that abstracts away much of the boilerplate required to build modern front-end applications, such as state management, content updates, or routing. We have extensive experience building React applications in the team, which we're looking to leverage. 

## Decision

We will use NextJS as the framework to build the front-end for the Bioacoustics project. 

## Options Considered

- **project-seed** is Development Seed's in-house project template. It bundles React and ui-seed, our component library, and uses Browserify for bundling JS. Project-seed uses some outdated technology, specifically Browserify, which prevents us from applying state-of-the-art architectural patterns, including code splitting. 
- Vanilla **ReactJS** focusses on building front-end applications but it's very bare-bones. Functionality for routing or code splitting needs to be added on top with the help of additional libraries. 
- **Remix** is a full-stack framework allowing you to build both the front-end and the back-end of a web application and integrate both. We already decided to build the backend using Python and Django and we won't use much of Remix' advantages. 
- **GatsbyJS** is for more classic websites, like developmentseed.org. While we can build a dynamic web app with it, we wouldn't use many of Gatsby's features, specifically using GraphQL to integrate with headless CMS. 

## Consequences

By choosing NextJS we can draw from the extensive experience we have in the team building React applications.

NextJS gives us advantages over plain ReactJS, including:

- Ability to set up separate pages, which results in better code-splitting and a smaller bundle size for individual,
- Routing and linking functionality out of the box, and 
- A battle-tested build framework.

**Comment by LanesGood** [comment]
This is great for documentation, thanks for adding @oliverroick 

cc @developmentseed/google-bioacoustics
